File: gmchol.hs
import Data.List (foldl')
import Data.Array
import Data.Maybe (fromMaybe)
import Control.Monad (forM_)

gmchol :: Array (Int, Int) Double -> Int -> Array (Int, Int) Double
gmchol a n = runSTUArray $ do
    r <- newArray ((0, 0), (n-1, n-1)) 0.0
    e <- newArray ((0, 0), (n-1, n-1)) 0.0

    let normA = maximum [sum [abs (a ! (i, j)) | i <- [0..n-1]] | j <- [0..n-1]]
        gamm = maximum [abs (a ! (i, i)) | i <- [0..n-1]]
        delta = max (epsilon * normA) epsilon

    forM_ [0..n-1] $ \j -> do
        let thetaJ = foldl' (\acc i -> let sumVal = sum [r ! (k, i) * r ! (k, j) | k <- [0..i-1]] in
                                          acc `max` (a ! (i, j) - sumVal)) 0.0 [0..n-1]
        forM_ [0..n-1] $ \i -> do
            let sumVal = sum [r ! (k, i) * r ! (k, j) | k <- [0..i-1]]
            writeArray r (i, j) ((a ! (i, j) - sumVal) / (r ! (i, i)))
            when (i > j) $ writeArray r (i, j) 0.0

        let sumVal = sum [r ! (k, j) * r ! (k, j) | k <- [0..j-1]]
            phiJ = a ! (j, j) - sumVal
            xiJ = if (j + 1) < n then maximum [abs (a ! (i, j)) | i <- [j+1..n-1]] else abs (a ! (n-1, j))
            betaJ = sqrt (max gamm (max (xiJ / fromIntegral n) epsilon))

        eVal <- readArray e (j, j)
        let eJ = if delta >= max (abs phiJ) ((thetaJ * thetaJ) / (betaJ * betaJ))
                 then delta - phiJ
                 else if abs phiJ >= max ((delta * delta) / (betaJ * betaJ)) delta
                      then abs phiJ - phiJ
                      else if ((thetaJ * thetaJ) / (betaJ * betaJ)) >= max delta (abs phiJ)
                           then ((thetaJ * thetaJ) / (betaJ * betaJ)) - phiJ
                           else 0.0
        writeArray e (j, j) eJ
        writeArray r (j, j) (sqrt (a ! (j, j) - sumVal + eJ))

    return r

epsilon :: Double
epsilon = 1.1102230246251565e-16

main :: IO ()
main = do
    let n = 3
        a = array ((0, 0), (n-1, n-1)) [((0, 0), 4), ((0, 1), 12), ((0, 2), -16),
                                          ((1, 0), 12), ((1, 1), 37), ((1, 2), -43),
                                          ((2, 0), -16), ((2, 1), -43), ((2, 2), 98)]
        r = gmchol a n

    putStrLn "R matrix:"
    forM_ [0..n-1] $ \i -> do
        forM_ [0..n-1] $ \j -> do
            putStr (show (r ! (i, j)) ++ " ")
        putStrLn ""

File: gmchol.rs
use std::f64;
use std::iter::repeat_with;

fn gmchol(A: &[Vec<f64>], n: usize) -> Vec<Vec<f64>> {
    let mut R = vec![vec![0.0; n]; n];
    let mut E = vec![vec![0.0; n]; n];

    let mut norm_A: f64 = 0.0;
    for j in 0..n {
        let col_sum: f64 = A.iter().map(|row| row[j].abs()).sum();
        norm_A = norm_A.max(col_sum);
    }

    let mut gamm: f64 = 0.0;
    for i in 0..n {
        gamm = gamm.max(A[i][i].abs());
    }

    let delta = f64::EPSILON.max(f64::EPSILON * norm_A);

    for j in 0..n {
        let mut theta_j = 0.0;

        for i in 0..n {
            let sum_val: f64 = (0..i).map(|k| R[k][i] * R[k][j]).sum();
            R[i][j] = (A[i][j] - sum_val) / R[i][i];

            if (A[i][j] - sum_val) > theta_j {
                theta_j = A[i][j] - sum_val;
            }

            if i > j {
                R[i][j] = 0.0;
            }
        }

        let sum_val: f64 = (0..j).map(|k| R[k][j] * R[k][j]).sum();
        let phi_j = A[j][j] - sum_val;

        let xi_j = if j + 1 < n {
            (j + 1..n).map(|i| A[i][j].abs()).fold(0.0, f64::max)
        } else {
            A[n - 1][j].abs()
        };

        let beta_j = (gamm.max(xi_j / n as f64)).sqrt();

        if delta >= phi_j.abs().max((theta_j * theta_j) / (beta_j * beta_j)) {
            E[j][j] = delta - phi_j;
        } else if phi_j.abs() >= (delta * delta) / (beta_j * beta_j).max(delta) {
            E[j][j] = phi_j.abs() - phi_j;
        } else if (theta_j * theta_j) / (beta_j * beta_j) >= delta.max(phi_j.abs()) {
            E[j][j] = (theta_j * theta_j) / (beta_j * beta_j) - phi_j;
        }

        R[j][j] = (A[j][j] - sum_val + E[j][j]).sqrt();
    }

    R
}

fn main() {
    let n = 3; // Matrix size
    let mut A = vec![
        vec![4.0, 12.0, -16.0],
        vec![12.0, 37.0, -43.0],
        vec![-16.0, -43.0, 98.0],
    ];

    let R = gmchol(&A, n);

    println!("R matrix:");
    for i in 0..n {
        for j in 0..n {
            print!("{} ", R[i][j]);
        }
        println!();
    }
}

